### PR TITLE
Better testing pyramid

### PR DESCRIPTION
**Pull request by charypar**

### CODE REVIEW

File: _posts/2022-07-13-in-defence-of-the-testing-pyramid.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+> The goal of each layer is to more cheaply and more regularly gain a reasonably high confidence that the layer above will pass, so that it does not need to be run as often
+
+In other words, the lower layers are an optimisation of executing the upper layers, in order to reduce their cost while maintaining a high enough overall confidence in the system.
+
+### Changing tests with implementation is fine, top-heavy pyramids are not fine
+
+Some will argue that pushing tests lower, closer to the implementation will lead to us having to regularly change tests when we change implementation, and that the execution time saved will instead be spent on keeping the lower layer tests up to date. To that, I will say that the idea that it should possible to change implementation without changing tests is just… nonsense.
```
Comment (gregdorward) [inline _posts/2022-07-13-in-defence-of-the-testing-pyramid.md:48]: it should possible (missing "be")

File: _posts/2022-07-13-in-defence-of-the-testing-pyramid.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+1. All layers of the pyramid are necessary in order to achieve a high confidence in your system.
+1. Always prefer testing functionality on lowest possible layer. Testing lower down will be harder, but worth the effort.
+1. When a layer of testing starts slowing you down, reduce the volume of that layer and replace it by coverage on the lower layers. Introduce new forms of testing on lower layers.
+1. Do not give in to the temptation of moving layers “left” in the development process, executing the more expensive ones earlier or more often.
+
+The last principle is worth talking about a little more. It suggests, for example, that running automated, system level, end to end test on each pull requests, as tempting as it sounds, is not worth the effort and complexity. End to end tests require an environment to run in (see assumptions), which would need to be created and destroyed (or allocated and cleaned up) for each pull request. In my experience, this is really complex, slow and highly unlikely to pay off in additional confidence gained, compared to running the end to end tests on the main branch, after merging the pull request.
```
Comment (gregdorward) [inline _posts/2022-07-13-in-defence-of-the-testing-pyramid.md:76]: end to end test(s)

File: _posts/2022-07-13-in-defence-of-the-testing-pyramid.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+## Built to be testable
+
+I have one final observation about all this: If, despite your best effort, you can’t seem to push test coverage down the pyramid, you may need to revisit your architecture, or even technology choices. Testability entirely depends of how your system is built.
+
+In a system without stubs of external dependencies, achieving reasonably high coverage with automated end to end tests will be difficult and slow. In a system with services sharing state (a database for example), pushing tests from layer 2 down to layer 1 will be difficult. In a codebase with no concept of dependency injection (even a really simple one), moving coverage from functional tests to unit tests will be difficult (and require extensive, messy mocking and stubbing). In a system without contract testing or shared API types ensuring client/server alignment on layer 0, a lot more layer 3 tests will be required to gain the same level of confidence. In a dynamic language with few static guarantees, layer 1 will almost certainly be much larger than layer 0.
```
Comment (gregdorward) [inline _posts/2022-07-13-in-defence-of-the-testing-pyramid.md:86]: Great summary of testability

**Comment by gregdorward** [comment]
Left a few comments - apologies that they're not inline, I was doing it on mobile. Really nice blog

**Comment by charypar** [comment]
@gregdorward Thanks for having a look!

File: _posts/2022-07-13-in-defence-of-the-testing-pyramid.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+- They are stateful
+
+All these reasons make repeatable, reliable testing solely against external systems difficult. But fully relying on mocks is not a reliable strategy either. We need to be sure that the mocks still behave like the actual system. We need to do both.
+
+## Guiding principles
+
+So, we now have a more complete version of the pyramid. To make it an actionable model, we just need some overall guiding principles:
+
+1. All layers of the pyramid are necessary in order to achieve a high confidence in your system.
+1. Always prefer testing functionality on lowest possible layer. Testing lower down will be harder, but worth the effort.
```
[the remark this answers is not in the document]
Comment (charypar) [reply _posts/2022-07-13-in-defence-of-the-testing-pyramid.md]: I'm probably thinking of poorly architected systems there. Nevermind, it's confusing, I've removed that second sentence.

**Pull request merged**